doc/tutorials/platformer/app/lib/actors/ember.dart

```diff
@@ -0,0 +1,166 @@
+import 'package:flame/collisions.dart';
+import 'package:flame/components.dart';
+import 'package:flame/effects.dart';
+import 'package:flutter/services.dart';
+
+import '../ember_quest.dart';
+import '../objects/ground_block.dart';
+import '../objects/platform_block.dart';
+import '../objects/star.dart';
+import 'water_enemy.dart';
+
+class EmberPlayer extends SpriteAnimationComponent
+    with KeyboardHandler, CollisionCallbacks, HasGameRef<EmberQuestGame> {
+  EmberPlayer({
+    required super.position,
+  }) : super(size: Vector2.all(64), anchor: Anchor.center);
+
+  final Vector2 velocity = Vector2.zero();
+  final Vector2 fromAbove = Vector2(0, -1);
+  final double gravity = 15;
+  final double jumpSpeed = 600;
+  final double moveSpeed = 200;
+  int horizontalDirection = 0;
+
+  bool hasJumped = false;
+  bool isOnGround = false;
+  bool hitByEnemy = false;
+
+  @override
+  Future<void> onLoad() async {
+    animation = SpriteAnimation.fromFrameData(
+      gameRef.images.fromCache('ember.png'),
+      SpriteAnimationData.sequenced(
+        amount: 4,
+        textureSize: Vector2.all(16),
+        stepTime: 0.12,
+      ),
+    );
+
+    add(
+      CircleHitbox(),
+    );
+  }
+
+  @override
+  bool onKeyEvent(RawKeyEvent event, Set<LogicalKeyboardKey> keysPressed) {
+    horizontalDirection = 0;
+    horizontalDirection += (keysPressed.contains(LogicalKeyboardKey.keyA) ||
+            keysPressed.contains(LogicalKeyboardKey.arrowLeft))
+        ? -1
+        : 0;
+    horizontalDirection += (keysPressed.contains(LogicalKeyboardKey.keyD) ||
+            keysPressed.contains(LogicalKeyboardKey.arrowRight))
+        ? 1
+        : 0;
+
+    hasJumped = keysPressed.contains(LogicalKeyboardKey.space);
+    return true;
+  }
+
+  @override
+  void update(double dt) {
+    velocity.x = horizontalDirection * moveSpeed;
+    gameRef.objectSpeed = 0;
+    // Prevent ember from going backwards at screen edge.
+    if (position.x - 36 <= 0 && horizontalDirection < 0) {
+      velocity.x = 0;
+    }
+    // Prevent ember from going beyond half screen.
+    if (position.x + 64 >= gameRef.size.x / 2 && horizontalDirection > 0) {
+      velocity.x = 0;
+      gameRef.objectSpeed = -moveSpeed;
+    }
+
+    // Apply basic gravity.
+    velocity.y += gravity;
+
+    // Determine if ember has jumped.
+    if (hasJumped) {
+      if (isOnGround) {
+        velocity.y = -jumpSpeed;
+        isOnGround = false;
+      }
+      hasJumped = false;
+    }
+
+    // Prevent ember from jumping to crazy fast.
+    velocity.y = velocity.y.clamp(-jumpSpeed, 150);
```

Comment:
In that case, does it need a new constant or could it simply just be clamped between - and + jumpSpeed?

Reply:
Done.  However, the clamp is `-jumpSpeed` because `jumpSpeed` is 600 but `terminalVelocity` falling is 150 so the collision isn't so fast it blows through the platform or the ground.